concent_api/core/models.py

```diff
@@ -73,3 +79,252 @@ class MessageAuth(Model):
+    reject_report_computed_task = OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_reject_report_computed_task')
+    subtask_results_accepted    = OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_subtask_results_accepted')
+    subtask_results_rejected    = OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_subtask_results_rejected')
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._current_state_name = None
+
+    @classmethod
+    def from_db(cls, db, field_names, values):
+        new = super().from_db(db, field_names, values)
+        new._current_state_name = new.state  # pylint: disable=no-member
+        return new
+
+    def clean(self):
+        super().clean()
+
+        # next_deadline can be None only for passive states
+        if not self._state.adding and self.next_deadline is None and self.state_enum not in self.PASSIVE_STATES:
+            raise ValidationError({
+                'next_deadline': 'next_deadline is None for active state.'
+            })
+
+        # next_deadline must be None in passive states
+        if not self._state.adding and self.next_deadline is not None and self.state_enum in self.PASSIVE_STATES:
+            raise ValidationError({
+                'next_deadline': 'next_deadline is not None for passive state.'
+            })
+
+        # State transition can happen only by defined rule
+        # but not when we create new object
+        # and not when state is not being changed
+        if (
+            not self._state.adding and
+            self._current_state_enum != self.state_enum and
+            self._current_state_enum not in self.POSSIBLE_TRANSITIONS_TO[self.state_enum]
+        ):
+            raise ValidationError({
+                'state': 'Subtask cannot change its state from {} to {}.'.format(
+                    self._current_state_name, self.state
+                )
+            })
+        else:
+            self._current_state_name = self.state
+
+        # Both ack_report_computed_task and reject_report_computed_task cannot set at the same time.
+        if self.ack_report_computed_task is not None and self.reject_report_computed_task is not None:
+            raise ValidationError(
+                'Both ack_report_computed_task and reject_report_computed_task cannot set at the same time.'
+            )
+
+        # Requestor and provider cannot be the same clients
+        if self.requestor_id == self.provider_id:
+            raise ValidationError('Requestor and provided are the same clients.')
+
+        # Check if all required related messages are not None in current state.
+        for stored_message_name, states in Subtask.REQUIRED_RELATED_MESSAGES_IN_STATES.items():
+            if self.state_enum in states and getattr(self, stored_message_name) is None:
+                raise ValidationError({
```

Comment:
You're only checking if it's `None` when if should be. You're not checking that it's not `None` when it shouldn't.

Reply:
done :+1: